src/engine/engineworkerscheduler.cpp

```diff
@@ -45,11 +48,13 @@ void EngineWorkerScheduler::runWorkers() {
 void EngineWorkerScheduler::run() {
     while (!m_bQuit) {
         Event::start("EngineWorkerScheduler");
-        EngineWorker* pWorker = NULL;
-        while (m_scheduleFIFO.read(&pWorker, 1) == 1) {
-            if (pWorker) {
-                pWorker->wake();
+        QHash<EngineWorker*, bool>::iterator i = m_workerStates.begin();
+        while (i != m_workerStates.end()) {
+            if (i.key() && i.value()) {
+                i.key()->wake();
+                i.value() = false;
```

Comment:
This is a typical race condition: The worker thread might re-schedule itself after woken up before the scheduler thread resets the flag to false. Even if the probability might be low, those are the critical spots that distinguish correct from wrong multi-threaded code!

Reply:
If you can ensure that the worker is not running during this loop then switching both lines would be sufficient. Without the "not running" assumption you need to use an atomic flag and clear() instead of a simple boolean. This is what I would prefer instead of relying on some implicit assumptions that may change at any time without notice.